File: poc-webmail/outils/bundle.py
#!/usr/bin/env python3
"""Produit une page autonome à partir des sources.

Le POC est multi-fichiers pour être maintenable ; il doit aussi pouvoir être
déposé quelque part en un seul fichier, ou ouvert sans serveur. Ce script lit
index.html, inline la feuille de style et concatène les scripts DANS L'ORDRE
QU'IL DÉCLARE — le même ordre que le harnais de test.

    python3 outils/bundle.py [destination.html]
"""
import io, os, re, sys, datetime

RACINE = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..")
SORTIE = sys.argv[1] if len(sys.argv) > 1 else os.path.join(RACINE, "dist", "index.html")

html = io.open(os.path.join(RACINE, "index.html"), encoding="utf-8").read()

# 1. la feuille de style
css = io.open(os.path.join(RACINE, "css", "app.css"), encoding="utf-8").read()
html = html.replace('<link rel="stylesheet" href="css/app.css">',
                    "<style>\n" + css + "\n</style>")

# 2. les scripts, dans l'ordre déclaré
fichiers = re.findall(r'<script src="([^"]+)"></script>', html)
if not fichiers:
    sys.exit("aucun script trouvé dans index.html")
morceaux = []
for f in fichiers:
    src = io.open(os.path.join(RACINE, f), encoding="utf-8").read()
    morceaux.append("/* ===== " + f + " ===== */\n" + src)
bloc = "<script>\n" + "\n".join(morceaux) + "\n</script>"

premier = '<script src="%s"></script>' % fichiers[0]
html = html.replace(premier, bloc, 1)
for f in fichiers[1:]:
    html = html.replace('<script src="%s"></script>\n' % f, "")
    html = html.replace('<script src="%s"></script>' % f, "")

# 3. commentaires d'inclusion devenus inutiles
html = re.sub(r"\n<!--[^>]*?-->\n(?=\s*(<script>|</body>))", "\n", html, flags=re.S)
html = re.sub(r"\n{3,}", "\n\n", html)

marque = ("<!-- Page autonome engendrée par outils/bundle.py le %s.\n"
          "     NE PAS ÉDITER : les sources sont dans js/ et css/. -->\n"
          % datetime.datetime.now().strftime("%Y-%m-%d %H:%M"))
html = html.replace("<!doctype html>", "<!doctype html>\n" + marque, 1)

os.makedirs(os.path.dirname(os.path.abspath(SORTIE)), exist_ok=True)
io.open(SORTIE, "w", encoding="utf-8").write(html)
print("%s : %d fichiers, %.0f Ko" % (os.path.relpath(SORTIE, RACINE),
                                     len(fichiers) + 1, len(html.encode()) / 1024))

File: webmail/test/bundle.js
/* Le bundle doit se comporter exactement comme les sources : même corpus, mêmes
   vues, même trace. Sinon le fichier déployé n'est pas ce qu'on a testé. */
"use strict";
const fs = require("fs"), path = require("path"), vm = require("vm");
const { creerDocument, creerStockage } = require("./fake-dom");
const { demarrer, eq, vrai, bilan } = require("./run");

const RACINE = path.join(__dirname, "..");
const BUNDLE = path.join(RACINE, "dist", "index.html");

if (!fs.existsSync(BUNDLE)) {
  console.log("dist/index.html absent — lancer : python3 outils/bundle.py");
  process.exit(1);
}
const html = fs.readFileSync(BUNDLE, "utf8");

console.log("— forme du bundle ————————————————————————————");
eq((html.match(/<script src=/g) || []).length, 0, "aucun script externe");
eq(/<link[^>]*href="css\/app\.css"/.test(html), false, "aucune feuille de style externe (la balise, pas la chaîne : le CDC embarqué la cite)");
vrai(html.includes("<style>"), "la feuille est inlinée");
vrai(html.includes("engendrée par outils/bundle.py"), "le fichier se déclare engendré");
vrai(html.length > 150000, "taille : " + Math.round(html.length / 1024) + " Ko");
vrai(html.includes("window.ABX_INLINE = true"), "la page autonome dit au chargeur que tout est inline (D157)");
vrai(html.includes("src/poc/pocbar.js =====") && html.includes("src/poc/serveur.poc.js ====="), "elle embarque le POC : poc/poc y marche hors ligne");

console.log("— il tourne ——————————————————————————————————");
/* Le bundle ne contient qu'UN bloc script, mais son contenu peut citer la
   chaîne « <script> » (le CDC parle d'affichage sûr) : on prend du premier
   <script> au DERNIER </script>, pas au premier venu. */
const js = html.slice(html.indexOf("<script>") + 8, html.lastIndexOf("</script>"));
const doc = creerDocument(), ls = creerStockage();
/* la maquette est une session (D157) : la page autonome s'ouvre avec poc/poc, tout est inline */
ls.setItem("abx.session", JSON.stringify({ mode: "poc", jeton: "poc", compte: "poc" }));
const sandbox = { console: { log(){}, warn(){}, error(){} }, document: doc, localStorage: ls,
  matchMedia: () => ({ matches:false }), addEventListener: () => {},
  alert: () => {}, confirm: () => true, location: { reload(){} },
  setTimeout, clearTimeout, encodeURIComponent, Math, Date, JSON };
sandbox.window = sandbox;
sandbox.Promise = Promise; sandbox.setImmediate = setImmediate;
vm.runInContext(js, vm.createContext(sandbox), { filename: "dist/index.html" });
const B = sandbox.ABX;
(async () => {
/* l'amorçage est asynchrone (D141) : on laisse la page se peindre avant de lire le DOM */
await new Promise(r => setImmediate(() => (B.pret || Promise.resolve()).then(() => setImmediate(r), () => setImmediate(r))));

console.log("— même comportement que les sources ————————————");
const src = demarrer(creerStockage()).ABX;
eq(B.Corpus.tous.length, src.Corpus.tous.length,
   "même corpus (" + B.Corpus.tous.length + " messages)");
eq(B.Registry.liste().length, src.Registry.liste().length, "mêmes vues partielles");
eq(Object.keys(B.Providers.CAPACITES).length, Object.keys(src.Providers.CAPACITES).length,
   "mêmes capacités");
eq(B.CDC.decisions.length, src.CDC.decisions.length, "même index du CDC");
vrai(doc.getElementById("nav").innerHTML.includes("Fournisseurs"), "l'arborescence est peinte");
vrai(doc.getElementById("list").innerHTML.includes("msg"), "la liste aussi");
vrai(doc.getElementById("qpanel").innerHTML.includes("casc"), "et la trace");
vrai(B.PocBar && B.PocBar.chargee && B.Session.simulee(), "la barre du POC est là, la session est simulée");

bilan();
})();
